File: Sources/App/Models/Album.swift
//
//  Album.swift
//  FinchServer
//
//  Created by Stefan Klein Nulent on 25/01/2025.
//

import Vapor
import Fluent
import SwiftGD

final class Album: Model, @unchecked Sendable {
    
    // MARK: - Constants
    
    static let schema = "albums"
    
    
    // MARK: - Private Properties
    
    @Field(key: "albumtypes")
    private var typesValue: String
    
    
    
    // MARK: - Properties
    
    @ID(custom: .id, generatedBy: .database)
    var id: Int?
    
    @Field(key: "albumartist")
    var albumArtist: String
    
    @Field(key: "albumartists_sort")
    var albumArtistSortKey: String
    
    @Field(key: "album")
    var title: String
    
    @Field(key: "release_group_title")
    var releaseGroupTitle: String
    
    @Field(key: "albumtype")
    var albumType: String

    @Field(key: "added")
    var addedAt: Date
    
    @Field(key: "genre")
    var genre: String?
    
    @Field(key: "style")
    var style: String?
    
    @Field(key: "year")
    var year: Int
    
    @Field(key: "disctotal")
    var discCount: Int
    
    @Field(key: "discogs_albumid")
    var discogsAlbumId: Int?
    
    @Field(key: "discogs_artistid")
    var discogsArtistId: Int?
    
    @Field(key: "discogs_labelid")
    var discogsLabelId: Int?
    
    @Field(key: "label")
    var label: String?
    
    @Field(key: "barcode")
    var barcode: String?
    
    @Field(key: "asin")
    var asin: String?
    
    @Field(key: "catalognum")
    var catalogNumber: String?
    
    @Field(key: "country")
    var country: String?
    
    @Field(key: "artpath")
    var artworkPathData: Data?
    
    @Children(for: \.$album)
    var attributes: [AlbumAttribute]
    
    @Children(for: \.$album)
    var items: [Item]
    
    var types: [AlbumType] {
        typesValue
            .split(separator: ";")
            .compactMap{ AlbumType(rawValue: $0.trimmingCharacters(in: .whitespaces)) }
    }
    
    var artworkPath: String? {
        get {
            guard isArtworkAvailable, let artworkPathData else { return nil }
            
            return String(data: artworkPathData, encoding: .utf8)
        }
        set {
            artworkPathData = newValue?.data(using: .utf8)
        }
    }
    
    var artworkURL: URL? {
        guard let artworkPath else { return nil }
        
        return URL(fileURLWithPath: artworkPath)
    }
    
    var artwork: Image? {
        guard let artworkURL else { return nil }
        
        return Image(url: artworkURL)
    }
    
    var isArtworkAvailable: Bool {
        if let artworkPathData, !artworkPathData.isEmpty {
            return true
        }
        
        return false
    }
    
    var artworkThumbnailFilename: String? {
        guard let discogsAlbumId, let fileExtension = artworkURL?.pathExtension else { return nil }
        
        return "album_\(discogsAlbumId)_thumbnail.\(fileExtension)"
    }
    
    
    
    // MARK: - Functions
    
    func artworkThumbnailPath(req: Request) -> String? {
        guard let artwork, let artworkThumbnailFilename else { return nil }
        
        let thumbnailPath = req.application.directory.thumbnailsDirectory + artworkThumbnailFilename
        
        if FileManager.default.fileExists(atPath: thumbnailPath) {
            return thumbnailPath
        }
        
        guard let thumbnail = artwork.resizedTo(width: 300) else { return nil }
        
        let thumbnailURL = URL(fileURLWithPath: thumbnailPath)
        thumbnail.write(to: thumbnailURL)
        
        return thumbnailPath
    }
}

File: FinchServer/Beets/Album.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Text;
using SixLabors.ImageSharp;
using SixLabors.ImageSharp.Processing;

namespace FinchServer.Beets;

[Table("albums")]
public class Album {
    
    // - Properties
    
    [Key]
    public int Id { get; set; }
    
    [Column("album")]
    public required string Title { get; set; }
    
    [Column("albumtype")]
    public required string AlbumType { get; set; }
    
    [Column("albumartist")]
    public required string AlbumArtist { get; set; }
    
    [Column("added")]
    public required double Added { get; set; }
    
    [Column("genre")]
    public string? Genre { get; set; }
    
    [Column("style")]
    public string? Style { get; set; }
    
    [Column("year")]
    public int Year { get; set; }
    
    [Column("disctotal")]
    public int DiscCount { get; set; }
    
    [Column("discogs_albumid")]
    public int? DiscogsAlbumId { get; set; }
    
    [Column("discogs_artistid")]
    public int? DiscogsArtistId { get; set; }
    
    [Column("discogs_labelid")]
    public int? DiscogsLabelId { get; set; }
    
    [Column("label")]
    public string? Label { get; set; }
    
    [Column("barcode")]
    public string? Barcode { get; set; }
    
    [Column("asin")]
    public string? Asin { get; set; }
    
    [Column("catalognum")]
    public string? CatalogNumber { get; set; }
    
    [Column("country")]
    public string? Country { get; set; }
    
    [Column("artpath")]
    public byte[]? ArtworkPathData { get; set; }
    
    public required ICollection<Item> Items { get; set; }

    public string? ArtworkPath => ArtworkPathData == null ? null : Encoding.UTF8.GetString(ArtworkPathData);
    public bool IsArtworkAvailable => ArtworkPathData != null && ArtworkPathData.Length > 0;

    public string? ArtworkThumbnailFilename {
        get {
            if (DiscogsAlbumId == null || ArtworkPath == null) return null;
            var extension = Path.GetExtension(ArtworkPath);
            
            return $"album_{DiscogsAlbumId}_thumbnail{extension}";
        }
    }
    
    
    // - Functions

    public async Task<string?> ArtworkThumbnailPath(IWebHostEnvironment environment) {
        if (ArtworkThumbnailFilename == null) return null;
        
        var thumbnailPath = Path.Combine(environment.ContentRootPath, "Resources/Thumbnails", ArtworkThumbnailFilename);
        if (File.Exists(thumbnailPath)) return thumbnailPath;
        
        if (ArtworkPath == null) return null;
        
        using var image = await Image.LoadAsync(ArtworkPath);
        image.Mutate(i => i.Resize(300, 0));
        await image.SaveAsync(thumbnailPath);

        return thumbnailPath;
    }
}